### ISSUE TITLE
Float.asCompileString doesn't `interpret` back to original Float

### ISSUE THREAD

**Issue by totalgee** [supercollider/supercollider#4800]
## Environment

* SuperCollider version: 3.10.4
* Operating system: Windows 10 Pro
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce

The Help for `Float.asCompileString` says:

> Returns:
> a string that when interpreted matches the receiver, if the number is within the range given in storeOn.
> 
> `a = 2.81882773638;`
> `a.asCompileString.interpret == a;`

But this is not true in general, for example:

```supercollider
a = 1/7; a.asCompileString.interpret == a
// -> false

pi.asCompileString.interpret == pi
// -> false

```

It also fails for `1/3`, `1/6` and many other numbers.

## Expected vs. actual behavior

Expected it to return true in the above cases (or else the documentation to be changed, whichever is the "correct" and desired behaviour.

Note that you can get the right answer now (workaround), by doing:

```supercollider
a = 1/7; a.asStringPrec(17).interpret == a
// -> true

pi.asStringPrec(17).interpret == pi
// -> true
```

**Comment by jamshark70** [comment]
What happens if we post floats with full precision is that "simple" values like 0.1 turn into 0.100000000001 and then we get floods of messages on the mailing list about how "numbers are broken." It isn't worth the heartache.

I'd accept this as a documentation bug -- the help that you quoted is flat wrong -- but I think we aren't going to change the string precision.

**Comment by brianlheim** [comment]
i think we should absolutely change this. this is what any reasonable user should expect from the documentation and intent of this method. knowing the behavior of floating point numbers is an unfortunate but necessary fact of life if you work with computers, so i don't see the benefit in hiding that from people when it makes sense not to.

**Comment by totalgee** [comment]
I agree, those long, incorrect-looking decimals are ugly. (Maybe use hex notation? Ha! No, since 0.1 is `0x1.999999999999ap-4` ;-)...

Is this maybe a question of `printOn` (current behaviour is good for Float, no horrible long things with extraneous decimals, such as @jamshark70 mentioned with 0.1) vs `storeOn`. If `storeOn` were overloaded for SC's Float type to return an "unambiguous representation that supports round-trips", that would ensure that for persistence purposes you'd be sure to get the same number back...which is the point, right? But the question is, how to do that? (for some numbers (`0.1`, `pi`), it is unambiguous to round to a maximum of 16 digits (`"0.1"`, `"3.141592653589793"`), for others (`1/6`, `1/7`), you need 17 decimals (`"0.16666666666666666"`, `"0.14285714285714285"`).

I guess this is a complex issue, and it seems there is a "best" solution, but it's not intuitive and it's not generally implemented correctly in C++ right now (`std::to_chars()` and `std::from_chars()` for float types *should* provide a shortest round-trip representation). See [this talk](https://youtu.be/4P_kbF0EbZM?t=804) (around 13 minutes in) for a great description, and how `std::to_chars()` (C++17 feature, but not really implemented yet for floats in most compilers) should be a solution once implemented. Or -- I think the library that implements this properly (the "Ryu" algorithm) is here: https://github.com/ulfjack/ryu

And, back to my original comment -- if this were implemented properly for round-trips, I guess we'd probably want the same implementation for `printOn` and `storeOn`, after all...

**Comment by brianlheim** [comment]
yes, i was thinking of to_chars and from_chars.. we could pull in the ryu project for this since it's C code but i don't think that's really worth it for this one bugfix. another totally unambiguous way to do it right now would be to use Float.from{32,64}Bits depending on the platform, but thats a bit ridiculous too.

i think we should document this as a known issue and then wait until we can use the standard library to solve it.

**Comment by totalgee** [comment]
I agree...just wanted to log it because I stumbled across this problem today. Not a rush, I'd use `std::to_/from_chars()`. I guess it's complex to solve, because this is one of the few C++17 things that still isn't really implemented by most of the big compilers/libraries (Microsoft allegedly now supports it as of VS 19.24, but libc++ and libstdc++ still only support integer types in `charconv`). (reference: "Elementary string conversions" from https://en.cppreference.com/w/cpp/compiler_support#cpp17)

**Comment by brianlheim** [comment]
> I personally will get bored with explaining floats repeatedly so I'll leave those questions to others.

we can also use such modern techniques as pointing users to a FAQ or even including an explanation in the documentation! :)
